helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java

```diff
@@ -0,0 +1,256 @@
+            "Failed to write the realm ZNRecord in addShardingKey()! Namespace: {}, Realm: {}, ShardingKey: {}",
+            _namespace, realm, shardingKey, e);
+        return false;
+      }
+      return true;
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean deleteShardingKey(String realm, String shardingKey) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      ZNRecord znRecord =
+          _zkClient.readData(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm, true);
+      if (znRecord == null || !znRecord
+          .getListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY)
+          .contains(shardingKey)) {
+        // This realm does not exist or shardingKey doesn't exist. Return true!
+        return true;
+      }
+      znRecord.getListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY)
+          .remove(shardingKey);
+      // Overwrite this ZNRecord with the sharding key removed
+      try {
+        _zkClient
+            .writeData(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm, znRecord);
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to write the data back in deleteShardingKey()! Namespace: {}, Realm: {}, ShardingKey: {}",
+            _namespace, realm, shardingKey, e);
+        return false;
+      }
+      return true;
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean setRoutingData(Map<String, List<String>> routingData) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      if (routingData == null) {
+        throw new IllegalArgumentException("routingData given is null!");
+      }
+
+      // Remove existing routing data
+      for (String zkRealm : _zkClient
+          .getChildren(MetadataStoreRoutingConstants.ROUTING_DATA_PATH)) {
+        if (!_zkClient.delete(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + zkRealm)) {
+          LOG.error(
+              "Failed to delete existing routing data in setRoutingData()! Namespace: {}, Realm: {}",
+              _namespace, zkRealm);
+          return false;
+        }
+      }
+
+      // For each ZkRealm, write the given routing data to ZooKeeper
+      for (Map.Entry<String, List<String>> routingDataEntry : routingData.entrySet()) {
+        String zkRealm = routingDataEntry.getKey();
+        List<String> shardingKeyList = routingDataEntry.getValue();
+
+        ZNRecord znRecord = new ZNRecord(zkRealm);
+        znRecord
+            .setListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY, shardingKeyList);
+        try {
+          if (!_zkClient.exists(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + zkRealm)) {
```

Comment:
try catch so no need to check if the path exists?

Reply:
I believe @narendly 's logic here is to not write on top of an existing node, therefore the if statement is necessary even with try catch. 
However, this check is unnecessary for another reason: this is a synchronized function and each loop of `routingData` guarantees a different `zkRealm`, therefore the node will never already exist here. Correct me if I'm wrong, please.

Reply:
It's not completely necessary, but it's better to be safe since exists() calls are relatively cheap. I'd rather have an exist() call here than risk getting a ZkNoNodeException.